### PR TITLE
Implement Reports

### PR DESCRIPTION
**Pull request by ahlem-jarrar**

### CODE REVIEW

File: mobile/src/main/java/rs/readahead/washington/mobile/domain/entity/reports/TellaReportServer.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        if (other == null) {
+            return false
+        }
+        if (this === other) {
+            return true
+        }
+        if (other !is TellaReportServer) {
+            return false
+        }
+        return id == other.id
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/domain/entity/reports/TellaReportServer.kt:27]: This way it is possible to have the same project on the same server as a different connection ("server"). So I guess we should compare project too.

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/entry/ReportsEntryViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+            reportApiId = reportApiId,
+            description = description,
+            status = EntityStatus.FINALIZED,
+            widgetMediaFiles = files ?: emptyList(),
+            formPartStatus = FormMediaFileStatus.NOT_SUBMITTED,
+            serverId = server.id
+        )
+    }
+
+    fun getSubmittedFormInstance(
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/entry/ReportsEntryViewModel.kt:246]: This doesn't seem to be used anywhere.

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/entry/ReportsEntryViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+import javax.inject.Inject
+
+@HiltViewModel
+class ReportsEntryViewModel @Inject constructor(
+    private val getReportsServersUseCase: GetReportsServersUseCase,
+    private val saveReportFormInstanceUseCase: SaveReportFormInstanceUseCase,
+    private val getReportsUseCase: GetReportsUseCase,
+    private val deleteReportUseCase: DeleteReportUseCase,
+    private val getReportBundleUseCase: GetReportBundleUseCase,
+    private val submitReportUseCase: SubmitReportUseCase,
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/entry/ReportsEntryViewModel.kt:39]: submitReportUseCase is not used anywhere

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/viewpagerfragments/SubmittedReportsFragment.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            },
+            getString(R.string.action_delete) + " \"" + instance.title + "\"?",
+            requireContext().resources.getString(R.string.Delete_Submitted_Report_Confirmation),
+            requireContext().getString(R.string.action_yes),
+            requireContext().getString(R.string.action_no)
+        )
+    }
+
+    private fun loadEntityInstance(reportFormInstance: ReportFormInstance) {
+        viewModel.getDraftBundle(reportFormInstance)
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/viewpagerfragments/SubmittedReportsFragment.kt:95]: We should rename this fun getDraftBundle for something like getReportBundle. In the end, it is used in many places, not just Drafts so the name is misleading.

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/viewpagerfragments/OutboxReportsFragment.kt [hunk 5]
```diff
@@ -0,0 +1,112 @@
+package rs.readahead.washington.mobile.views.fragment.reports.viewpagerfragments
+
+import android.os.Bundle
+import android.view.View
+import androidx.fragment.app.viewModels
+import androidx.navigation.fragment.NavHostFragment
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/viewpagerfragments/OutboxReportsFragment.kt:6]: This is unused too

File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/entry/ReportsEntryViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+import rs.readahead.washington.mobile.domain.usecases.reports.*
+import rs.readahead.washington.mobile.util.fromJsonToObjectList
+import rs.readahead.washington.mobile.util.jobs.TellaRequestUploadJob
+import rs.readahead.washington.mobile.views.fragment.reports.adapter.ViewEntityTemplateItem
+import rs.readahead.washington.mobile.views.fragment.reports.mappers.toViewEntityInstanceItem
+import timber.log.Timber
+import javax.inject.Inject
+
+@HiltViewModel
+class ReportsEntryViewModel @Inject constructor(
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/reports/entry/ReportsEntryViewModel.kt:33]: This is our only ViewModel regarding Reports so maybe we should remove Entry from the name.

File: mobile/src/main/java/rs/readahead/washington/mobile/views/adapters/reports/ReportsFilesRecyclerViewAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+            Glide.with(filePreviewImg.context)
+                .using(glideLoader)
+                .load(VaultFileLoaderModel(vaultFile, VaultFileLoaderModel.LoadType.THUMBNAIL))
+                .diskCacheStrategy(DiskCacheStrategy.NONE)
+                .skipMemoryCache(true)
+                .into(filePreviewImg)
+        }
+
+        fun showAddLink() {
+            filePreviewImg.setBackground(context.getDrawable(R.drawable.transparent_solid))
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/adapters/reports/ReportsFilesRecyclerViewAdapter.kt:171]: This getDrawable requires API level 21, but we still support level 19+.

**Review by wafalakhal — APPROVED** [review]

**Pull request merged**